### PR TITLE
AUR packaging

### PR DESCRIPTION
**Pull request by ghbrown**
Adds packaging and build/test instructions for the Arch (Linux) User Repository (AUR), as per [the discourse discussion](https://chapel.discourse.group/t/when-is-setting-chpl-lib-pic-necessary-for-working-build/12339/21).

Additions:
- The chapel package for AUR tracks the static 1.26 release, but is non-functional due to issues in 1.26 that cause build issues on Arch. These issues have been resolved on `main`, but the chapel AUR package will be left broken until 1.27 releases (the online description of the broken AUR points to functioning chapel-git package).
- The chapel-git package for AUR tracks the main branch of the chapel repo and is fully functional.

Outstanding issues:
- updating chapel AUR to use 1.27 upon its release
- fixing the `git describe --long --tags --always` behavior for the main branch (currently main reports as version 1.24, while it is really 1.26 at the time of writing).

Possible future work:
- automated copying of packaging files from their remote sources (so maintainers don't need to manually copy and make PRs, etc), hopefully such a tool would work for many different packaging formats, OSes, etc. and not be a one-off for Arch
- testing of packaging; it is important to test both successful install **and** a functioning compiler **after** build and install, as we ran into multiple situations where the compiler built and checked, but did not actually work (#19913).

### CODE REVIEW

**Comment by ghbrown** [comment]
@mppf I also ran a build from source with `--prefix=/usr` that produces a working compiler. However, I noticed that one step of `make check` is to compile hello6 to an executable in `~/.chpl`. However, this hidden directory is not around after `make check`, and even when I create said directory (in case there was an error in that step) I can't find this file anywhere in `~/`.
Not directly related to this PR, but it came up when debugging, so I thought I'd see if this was expected behavior.

File: util/packaging/aur/chapel/.SRCINFO
Comment (mppf) [reply util/packaging/aur/chapel/.SRCINFO:13]: Is there a GMP package? If we can depend on a GMP package then we can set CHPL_GMP=system and save some build time.

**Comment by mppf** [comment]
I'm trying to debug the problem on Arch and the instructions you added have been very useful to me. I'm running `chpl hello.chpl --print-commands` with the installed version (that does not work) and compared to a normal source build the `-lhwloc` line seems to be missing - even though we do have `-L` paths for hwloc. In fact, if I compile with `chpl hello.chpl  --print-commands -lhwloc -lqthread`, then I get a program that compiles and runs. So, the question is, why are these arguments missing?

File: util/packaging/aur/chapel/PKGBUILD
Comment (ghbrown) [reply util/packaging/aur/chapel/PKGBUILD:10]: Great! I think I can add  `arm` to that array then.

File: util/packaging/aur/chapel/PKGBUILD
Comment (ghbrown) [reply util/packaging/aur/chapel/PKGBUILD:21]: `configure` creates `configured-prefix` and `chplconfig`? So instead of an environment variable is there a `configure` flag that would be preferable?

File: util/packaging/aur/chapel/PKGBUILD
Comment (ghbrown) [reply util/packaging/aur/chapel/PKGBUILD:28]: Yes, that's my understanding as well, and I think that could be the case.

It is possible for me to include a post install script, but I'm not sure how integrated it is in the build and install process, since the package will already be installed. The one example I have hear is people using this script to bring up end user license agreement and so on, but I'm sure we can get it do more. What exactly were you thinking would go in here?

File: util/packaging/aur/readme.md
Comment (ghbrown) [reply util/packaging/aur/readme.md:2]: Sure, I'd be happy to. The first bullet point is pretty simple I think, and for the second one I can write up a very simple overview of how to update and link to some useful resources.

**Comment by mppf** [comment]
I found a problem when `.la` files are removed in the install process. We had a fallback but it wasn't working. PR #19846 should fix it.

**Comment by ghbrown** [comment]
Thanks for answering the questions I raised a few days ago. I agree it doesn't make sense to keep a bunch of test-compile executables around. As for the `$TEMP` I suppose I saw it as a possible environment variable namespace issue, and didn't expect to see anything being edited or changed in "user-land", but since you're deleting everything anyway I suppose it could be seen as a feature to hook into the users `$TEMP`.

I'm glad you found the instructions useful and have narrowed down the issue! I'm not sure why these arguments aren't added automatically, but if I can dig into anything on my side or run any tests just let me know. I can look into them after the packaging and readme improvements from your comments.

**Comment by ghbrown** [comment]
The old AUR packaging used to have a dependence on `libtool` that I removed since you didn't include it in the installs of the Arch packages you depended on to build Chapel in your Arch VM.
I'm pretty sure removing this dependence isn't what's causing the issue (since we can both follow the `PKGBUILD`'s instructions without `makepkg` and get working compilers, but should I put back the dependence on `libtool`?

Great, glad you found it! Is there a way for me to get around this at build-time when packaging 1.26.0? Your PR will fix things automatically for `chapel-git` (which we'll get up and running soon after this PR merges, I hope), but for the chapel package I'll have to do the fixes after downloading the release. I suppose I could do patches, but is there a quicker and simpler way via configure and make flags, etc.?

**Comment by mppf** [comment]
> I suppose I could do patches, but is there a quicker and simpler way via configure and make flags, etc.?

I think it has to be a patch, for this one, unfortunately. The only thing I can think of is somehow getting the pkgbuild stuff to install the .la files but presumably there is a reason it does not.

File: util/packaging/aur/chapel/PKGBUILD
Comment (mppf) [reply util/packaging/aur/chapel/PKGBUILD:21]: Oh the environment variable is fine. `configure` saves the Chapel-related environment variables to `chplconfig`.

File: util/packaging/aur/chapel/PKGBUILD
Comment (mppf) [reply util/packaging/aur/chapel/PKGBUILD:28]: I was just thinking it would be nice to have some scripting so that the maintainer of the Chapel AUR package can easily check that the installation is working. I.e. it could just be a script that you know how to run after installing. If we have that, we can (eventually) get some nightly testing config to check it as well.

I don't know whether or not it would be reasonable to do this with an AUR post install script but I suspect you wouldn't want people just installing the package to always run it. (So then not a post-install).

**Comment by mppf** [comment]
> should I put back the dependence on `libtool`?

Would you be able to do an experiment to see if `libqthread.la` is installed if the package has a dependence on libtool? (It might be that makepkg is doing something more clever than we first expected, here).

I don't think we need `libtool` for any other reason, but could be forgetting/missing something.

File: util/packaging/aur/chapel/PKGBUILD
Comment (ghbrown) [reply util/packaging/aur/chapel/PKGBUILD:28]: I agree post-install is probably not the right tool here, but something like you describe would be nice.

I think the typical approach for AUR packages is to do verification with `make check` before install, but that didn't work for us. I'm sure there are issues with this approach, but the way I was going to verify that the installation is working is just by building and running it myself (then addressing user issues as they appear in comments on AUR). This is less of a problem for `chapel` since it's frozen until the next release, but perhaps would be problematic for `chapel-git` which changes (and possibly breaks) with every PR.

**Comment by ghbrown** [comment]
I'm relatively inexperienced at patching manually, but I'm not optimistic about it in this case.
First, just patching the files changed in PR #19846 gives build errors because I assume other files also changed in a way that breaks things. Chasing down all of these dependencies and patching them seems like a daunting task.
On the other hand, completely patching the 1.26 release with your branch produces an error-free build (not an error free compiler though!) and includes an almost 200MB patch file. That's larger than I'm comfortable putting in the AUR `chapel` repo, and I assume it wouldn't be welcome in this repo either.

Now onto the compile-time errors (though you are apparently not getting any?). Both the unpatched release with the compiler flags you suggested and the fully patched release with no compiler flags give errors like this
```
/usr/bin/ld: /usr//lib/chapel/1.26/third-party/hwloc/install/linux64-x86_64-native-llvm-pic-flat/lib/libhwloc.a(topology-linux.o): in function `hwloc_linux_set_thisthread_membind':
(.text+0x112e): undefined reference to `set_mempolicy'
/usr/bin/ld: (.text+0x1166): undefined reference to `set_mempolicy'
/usr/bin/ld: (.text+0x1200): undefined reference to `migrate_pages'
/usr/bin/ld: /usr//lib/chapel/1.26/third-party/hwloc/install/linux64-x86_64-native-llvm-pic-flat/lib/libhwloc.a(topology-linux.o): in function `hwloc_linux_get_thisthread_membind':
(.text+0x12b3): undefined reference to `get_mempolicy'
/usr/bin/ld: (.text+0x1362): undefined reference to `get_mempolicy'
/usr/bin/ld: /usr//lib/chapel/1.26/third-party/hwloc/install/linux64-x86_64-native-llvm-pic-flat/lib/libhwloc.a(topology-linux.o): in function `hwloc_linux_get_area_membind':
(.text+0x1506): undefined reference to `get_mempolicy'
/usr/bin/ld: (.text+0x16a2): undefined reference to `get_mempolicy'
/usr/bin/ld: (.text+0x17e9): undefined reference to `get_mempolicy'
/usr/bin/ld: /usr//lib/chapel/1.26/third-party/hwloc/install/linux64-x86_64-native-llvm-pic-flat/lib/libhwloc.a(topology-linux.o): in function `hwloc_linux_set_area_membind':
(.text+0x1b09): undefined reference to `mbind'
/usr/bin/ld: (.text+0x1b43): undefined reference to `mbind'
/usr/bin/ld: /usr//lib/chapel/1.26/third-party/hwloc/install/linux64-x86_64-native-llvm-pic-flat/lib/libhwloc.a(topology-linux.o): in function `hwloc_linux_get_area_memlocation':
(.text+0x1cad): undefined reference to `move_pages'
clang-13: error: linker command failed with exit code 1 (use -v to see invocation)
error: Make Binary - Linking
```
Perhaps I'm doing something wrong?

As for package dependencies, I added `GMP` and `libtool` for now. I tried to do some A/B testing to see if it made a difference, but wasn't sure exactly what to look for (please let me know and I'll retry).  For the `libtool` test I ran `# find /usr/* -name libqthread.la` after install, but got no matches.

**Comment by ghbrown** [comment]
@mppf I started work on a `chapel-git` package for AUR since I thought it might be nice to have a working package while we work on getting the main one shored up via patches, etc. However, even cloning and building from your unmerged `adjust-install` PR branch appears to result in the same errors as my previous comment.

I also tried to spot differences in compilation commands with `--print-commands`. The logfiles for the failing `chapel-git` and working version from your `adjust-install` branch (which I just manually configured and built) are attached, where I tried to compile `hello2`.
[chapel-git.log](https://github.com/chapel-lang/chapel/files/8776321/chapel-git.log)
[chapel_mppf.log](https://github.com/chapel-lang/chapel/files/8776322/chapel_mppf.log)
I don't know exactly what to look for, but a search for `hwloc` reveals that the working version has two more instances (8 compared to 6) than the failing version. However, I think the extra pair is just a duplicate of an already existing `-L -l` pair, so this may be a red herring.

**Comment by mppf** [comment]
The symbols you mentioned ought to be brought in by `-lnuma` which I would have imagined would be required by hwloc. Not sure what is going wrong though.

**Comment by mppf** [comment]
Oh, maybe the difference between our two systems is that you have libnuma installed somewhere and hwloc is auto-detecting it? E.g., for me, I get this:

```
$ gcc test.c -lnuma
/usr/bin/ld: cannot find -lnuma: No such file or directory
collect2: error: ld returned 1 exit status
```

I think I can install numactl with `pacman -S numactl`.
If I do that, `third-party/hwloc/install/linux64-x86_64-native-llvm-none-flat/lib/libhwloc.la` includes `dependency_libs=' -lm -lnuma'` and our build system normally figures out to add the `-lnuma` by looking at this `.la` file. But the Arch install process likes to delete the `.la` files, for some reason.

Given that the `-lnuma` shouldn't be added if hwloc didn't configure itself to work with libnuma, I don't think we can just always add `-lnuma`. However you could make the AUR package depend on `numactl` and then the AUR package could always add `-lnuma` (but I think that would have to be through a patch or some sort of custom setting to activate it).

> I'm relatively inexperienced at patching manually, but I'm not optimistic about it in this case. First, just patching the files changed in PR https://github.com/chapel-lang/chapel/pull/19846 gives build errors because I assume other files also changed in a way that breaks things.

I could probably create a series of patches against 1.26 but let's figure out how to get `main` working first.

**Comment by mppf** [comment]
It looks like (from my previous AUR install) that it does install the `hwloc.pc` file. Given that, I think that we could use it instead of `libhwloc.la` - potentially in all cases - but we could at least do that if `libhwloc.la` is missing.

An aside, I am curious why the AUR install process deletes the .la files, if you are able to figure that out.

**Comment by mppf** [comment]
@ghbrown - quick question - is `chapel-git` working for you at this point?

**Comment by ghbrown** [comment]
Yes, but I am working on my personal machine. Let me know if you find any issues.

File: util/packaging/aur/chapel-git/PKGBUILD
Comment (mppf) [inline util/packaging/aur/chapel-git/PKGBUILD:18]: I don't think this was using my branch, yet. However I am expecting to merge PR #19913 on Monday and my branch is working in my own testing.

Comment (mppf) [reply util/packaging/aur/chapel-git/PKGBUILD:18]: (my `main` branch usually just tracks upstream/main but I just pushed a merge of PR #19913 in to it, FWIW).

Comment (ghbrown) [reply util/packaging/aur/chapel-git/PKGBUILD:18]: Right, I check out your `less-la` branch on line 28, if that's what you are referring to. Hopefully that's doing what I think it is, butI didn't have any issues.

Comment (mppf) [reply util/packaging/aur/chapel-git/PKGBUILD:18]: Ah, gotcha, thanks, I missed that.

**Comment by mppf** [comment]
Hi @ghbrown - I've just merged #19913, so when you are able it's a good time to update the AUR package with the `main` branch and check that it works. Once we have those things working OK I'll give this PR a real review and we'll seek to merge it.

**Comment by ghbrown** [comment]
Great, thanks. I can make the change tomorrow morning, although it will only be to `chapel-git` while this PR now contains both non-git and git versions.

**Comment by ghbrown** [comment]
@mppf When I use `git describe --long --tags --always` on `main` I get `1.24.0-7648-gbcc4ac3109`, but I expected it to begin with 1.26 or 1.27.
When I was building based on your branch there was no version number at the beginning.
This is my first time using `git describe`, but this output doesn't seem quite right to me. Perhaps I'm missing something?

**Comment by mppf** [comment]
> @mppf When I use `git describe --long --tags --always` on `main` I get `1.24.0-7648-gbcc4ac3109`, but I expected it to begin with 1.26 or 1.27. When I was building based on your branch there was no version number at the beginning. This is my first time using `git describe`, but this output doesn't seem quite right to me. Perhaps I'm missing something?

Huh, that is weird. As you are observing, `git describe --long --tags --always` worked with my `main` branch but it does happen with upstream:

```
$ cd /tmp
$ git clone https://github.com/chapel-lang/chapel
$ cd chapel
$ git describe --long --tags --always
1.24.0-7648-gbcc4ac3109
```

The tag `1.26.0` corresponds to commit d80a12706665c1aad8c2aae49aceb9fbd24b3302 but that commit is not in the `git log` history for the main branch. But AFAIK that is also true of my branch. So I don't know what's going on here.

You could consider doing some scripting to parse `compiler/main/version_num.h` or you could build `chpl` and use `chpl --version`.

**Comment by ghbrown** [comment]
Thanks for confirming and checking it out.
I think I need to do this before build, so I'll go for the scripting to combine the stored text with the hash at the end of `git describe`. Hopefully that will only be a temporary solution though.

**Comment by ghbrown** [comment]
I think I'm just going to replace the leading part of `git describe` with a hardcoded version or a version from the header. I can split the `git describe` on the first hypen and use the rest.

**Review by mppf — APPROVED** [review]
Please let me know if you feel this version is ready to merge into the Chapel repo or if there is something else you want to do first. I think it's ready to merge once you are satisfied with it. Thanks again for your efforts on the Chapel AUR package!

File: util/packaging/aur/readme.md
Comment (ghbrown) [reply util/packaging/aur/readme.md:15]: Yes, it's generated. The reason I included it is that the .SRCINFO must be generated on the maintainer's machine, such that the remote host and users who clone it only copy that file. Therefore, I must provide generate and include this file for all commits to `chapel-git`, etc. It's apparently related to security risks associated with parsing the `PKGBUILD` on the user machines (forget where I heard it, but [the Arch Wiki](https://wiki.archlinux.org/title/.SRCINFO) alludes to general problems.
I have no opinion on whether or not it should be included here though, so I'll let you decide.

Comment (mppf) [reply util/packaging/aur/readme.md:15]: Let's leave it in here in that case, thanks for some explanation on it.

File: util/packaging/aur/chapel-git/.SRCINFO
Comment (ghbrown) [reply util/packaging/aur/chapel-git/.SRCINFO:3]: Unsure exactly what you mean by "copied" it from something else?
The output of the pkgver() function automatically overwrites the pkgver line in PKGBUILD (I think the `makepkg` command probably feeds it to `sed` or something behind the scenes?).
As for what it means, I'm unsure on the central 4-digit part, but the very last part is `g<beginning of most recent commit hash`.

Comment (mppf) [reply util/packaging/aur/chapel-git/.SRCINFO:3]: When I made a hacky chapel-git for my own testing I just copied the version number from the other AUR project I was basing it on. Anyway since it will be overwritten in the PKGBUILD file by running `pkgver()`, it might be nice to have a comment to that effect next to the `pkgver=` line.

**Comment by ghbrown** [comment]
Overall, I'm happy with the state of the PR, but just to make sure we are on the same page about its outstanding deficiencies:
- the `chapel` AUR package is currently broken since it is based directly on 1.26, which has issues that prevent it from building on Arch; its description will point to `chapel-git` until 1.27
- the AUR version is being generated in a robust (but non-ideal) way, since `git describe` gives confusing outputs involving older versions at present; ideally this will be resolved and we can use more traditional commands in `pkgver()` in the future

**Comment by mppf** [comment]
@ghbrown - I forgot to ask earlier, but the PR message is now out of date. Would you mind updating it? It would be a good place to call out those deficiencies you just mentioned.

**Comment by ghbrown** [comment]
I hope that was the correct approach?

**Comment by mppf** [comment]
Oh I'm not sure what you did. I was just talking about the text box at the top of the PR that starts with

> As per [the discourse discussion](https://chapel.discourse.group/t/when-is-setting-chpl-lib-pic-necessary-for-working-build/12339/21), I provide AUR packaging files and instructions to test the build.

We (as a matter of practice) turn that into the commit message for the merge when we merge a PR, so its content matters. (Besides browsing PRs when understanding history).

Anyway this line is no longer true, right?

> Currently, the included recipe results in a no-error build, but the compiler gives linker errors when `chpl` is used on even simple programs.

**Comment by ghbrown** [comment]
Ah, thanks. I changed the final git commit message to be more substantial and detailed (following the trend of some others I'd seen in the git log), but didn't change the PR comment here.
To be clear, I should update and correct the first comment in this PR, then use that as the git commit message? Should I use markdown syntaxed version or just plain text?

**Comment by mppf** [comment]
> Ah, thanks. I changed the final git commit message to be more substantial and detailed (following the trend of some others I'd seen in the git log), but didn't change the PR comment here.

The commit messages can be whatever they were at the time - no need to update them all. The detailed commit messages you are seeing are from merge commits when we paste the PR message (that appears at the top of the PR webpage) into the merge commit message. For the most part our project focuses on PRs as the unit of archeology. 

> To be clear, I should update and correct the first comment in this PR, then use that as the git commit message? Should I use markdown syntaxed version or just plain text?

No need to worry about the git commit messages for this. Just update the PR message (which I think you are calling the 1st comment in the PR but it's at the top of the PR webpage either way). Markdown or plain text are both fine - up to you. I do tend to use markdown myself most of the time.

**Comment by ghbrown** [comment]
>The commit messages can be whatever they were at the time - no need to update them all. The detailed commit messages you are seeing are from merge commits when we paste the PR message

Yes, thankfully I only changed the last PR message (which was probably for the best anyway). The PR message should be changed too.

**Comment by ghbrown** [comment]
@mppf Thank your for all your help and support on this, it was a pleasure!

**Pull request merged**
